bundle/config/mutator/resource_key.go

```diff
@@ -0,0 +1,106 @@
+package mutator
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/databricks/cli/libs/dyn"
+)
+
+// ResourceKey uniquely identifies a resource in configuration.
+type ResourceKey struct {
+	// Type is the type of the resource. E.g. "jobs", "dashboards", etc.
+	Type string
+
+	// Path is the path to the resource.
+	Name string
+}
+
+// ResourceKeySet is a set of resource keys in configuration.
+type ResourceKeySet map[string]map[string]struct{}
+
+func NewResourceKeySet() ResourceKeySet {
+	return make(map[string]map[string]struct{})
+}
+
+// Add adds a resource key to the set.
+func (r ResourceKeySet) AddResourceKey(key ResourceKey) {
+	if _, ok := r[key.Type]; !ok {
+		r[key.Type] = make(map[string]struct{})
+	}
+
+	r[key.Type][key.Name] = struct{}{}
+}
+
+// Add adds resources associated with prefix/value to the set.
+//
+// prefix is a path to value in configuration.
+//
+// If prefix is a resource key ("resources.jobs.job_1"), it's added to the set.
+// If prefix is a resource value ("resources.jobs.job_1.foo"), resource key is extracted and added.
+// If prefix is a sub-tree ("resources" or "resources.jobs"), all resource keys are added.
+func (r ResourceKeySet) Add(prefix dyn.Path, value dyn.Value) error {
```

Comment:
It's hard to see why would this set support cutting unnecessary paths rather than working with resourceKey only.

Looking at how it is used, it maybe because callers do more work than necessary?

https://github.com/databricks/cli/pull/2590/files#r2022375624

Reply:
Agree. I've changed the API to AddPattern instead. It looks simpler now.